feat(cli): --filter flag

gvilums opened the pull request:
### What does this PR do?

Adds support for the `--filter <PATTERN>` argument for `bun run`. See the [pnpm docs](https://pnpm.io/filtering#--filter-package_name) for details.

Note that currently only `--filter <name_pattern>` and `--filter ./<dir_pattern>` are supported. Selecting dependencies or dependents of packages is not yet possible.

Resolves #2232
Resolves #5207 


### How did you verify your code works?
Tests in `test/cli/run/filter-workspace.test.ts`.

dylan-conway commented on src/cli/run_command.zig:
```diff
@@ -1010,8 +1009,61 @@ pub const RunCommand = struct {
+        defer {
+            // change back to the original directory once we're done
+            fsinstance.top_level_dir = olddir;
+            std.os.chdir(olddir) catch |err| {
```
```suggestion
            switch (bun.sys.chdir(olddir)) {
                .err => |err| {
                    ...
                },
                .success => {},
            }
```

dylan-conway commented on src/cli/run_command.zig:
```diff
@@ -1010,8 +1009,61 @@ pub const RunCommand = struct {
+        var ok = true;
+        for (workspace_paths.items) |path| {
+            Output.prettyErrorln("<d><b>In <r><yellow><d>{s}<r>:", .{path});
+            std.os.chdir(path) catch |err| {
```
`bun.sys.chdir` here too

dylan-conway commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,183 @@
+
+    while (true) : (workdir = std.fs.path.dirname(workdir) orelse break) {
+        const parent_trimmed = strings.withoutTrailingSlash(workdir);
+        var buf2: [bun.MAX_PATH_BYTES + 1]u8 = undefined;
```
```suggestion
        var buf2: bun.PathBuffer = undefined;
```
is the `+ 1` needed? null byte should be included

gvilums replied:
updated to use PathBuffer

dylan-conway commented on src/cli.zig:
```diff
@@ -402,20 +409,26 @@ pub const Arguments = struct {
+                Output.prettyErrorln("error resolving --cwd: {s}", .{@errorName(err)});
+                Global.exit(1);
+            };
+            std.os.chdir(out) catch |err| {
```
also `bun.sys.chdir`

dylan-conway commented on test/cli/run/filter-workspace.test.ts:
```diff
@@ -0,0 +1,145 @@
+    cmd.push("--filter", pattern);
+  }
+  cmd.push("present");
+  console.log(cmd);
```
these console logs can be removed?

gvilums replied:
removed

dylan-conway reviewed (commented):
Looks good so far. One thing that stands out as a difference from pnpm: `pnpm run --filter="*" debug` will run all `debug` scripts including one from the root package

gvilums commented:
Just pushed a commit that includes the root package in script resolution

dylan-conway commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,195 @@
+
+    // if we were not able to find a workspace root, try globbing for package.json files
+
+    var walker = Glob.BunGlobWalker{};
```
this glob walker should ignore `node_modules` and other directories that pnpm ignores if they exist. I believe this is done through the `ignore_filter_fn`

gvilums replied:
fixed in latest commit

dylan-conway commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,195 @@
+        }
+        const entry = Package.WorkspaceMap.Entry{ .name = try allocator.dupe(u8, parser.found_name), .version = null, .name_loc = bun.logger.Loc.Empty };
+        const dirpath = std.fs.path.dirname(path) orelse continue;
+        try workspace_map.insert(try allocator.dupe(u8, dirpath), entry);
```
package jsons found here may or may not actually be workspace packages. If the script isn't found in the `package.json` we shouldn't print an error
<img width="447" alt="Screenshot 2024-01-17 at 4 06 37 PM" src="https://github.com/oven-sh/bun/assets/35280289/cd3be29a-5843-487a-8c39-b61e04164301">

gvilums replied:
The `--if-present` flag accomplishes this, but maybe this should be the default, just like for pnpm.

Jarred-Sumner commented:
Running `bun-debug run --filter="*" echo hi` crashes

<img width="714" alt="image" src="https://github.com/oven-sh/bun/assets/709451/e092e490-b4cb-4ba5-8f18-f82feba99267">

Jarred-Sumner commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,195 @@
+        // add the root package to the workspace map, too
+        const entry = Package.WorkspaceMap.Entry{ .name = try allocator.dupe(u8, name), .version = null, .name_loc = bun.logger.Loc.Empty };
+        try workspace_map.insert(try allocator.dupe(u8, parent_trimmed), entry);
+
```
This duplicates memory twice.

gvilums replied:
Does it? From my understanding we dupe the name (part of the hash map value) in the first line, and duple the path (hash map key) in the second line.

Jarred-Sumner replied:
Right here:

https://github.com/oven-sh/bun/blob/52894121ea2436481b97b5318d1ce64176b18c0e/src/install/lockfile.zig#L3562-L3573

Jarred-Sumner commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,195 @@
+
+const Package = @import("../install/lockfile.zig").Package;
+
+fn findWorkspaceMembers(allocator: std.mem.Allocator, log: *bun.logger.Log, workspace_map: *Package.WorkspaceMap, workdir_: []const u8) !void {
```
In large directory trees, this approach takes seconds to show any output. We cannot ship it without being an iterator.

gvilums replied:
understood, will get on that 👍

Jarred-Sumner reviewed (changes requested):
4 issues before this can be shipped.

1) This approach is not fast enough. It needs to be an iterator so that the developer gets immediate feedback instead of waiting for potentially their entire directory tree to parse every package.json file.

2) There is a resource leak. After enough scripts are run, spawning any process fails with `SystemResources` which usually means a file descriptor leak.

3) it scans node_modules. It should not scan node_modules, .git, or any `.` directories.

4) We need tests for what happens when you run multiple JavaScript/TypeScript files in different projects same process (`bun run --filter="*" index.js`). Currently, Bun assumes only one JavaScript file every executes for a given thread. That assumption is not true when using these code paths.

5) It would be best to use the module resolution cache so that when running a JavaScript file later, the subsequent calls to getdents64() get cached. This will involve a small change to the GlobWalker.

Jarred-Sumner commented on src/cli/run_command.zig:
```diff
@@ -1142,6 +1147,8 @@ pub const RunCommand = struct {
                 .result => {},
             }
             fsinstance.top_level_dir = path;
```
only this one should be necessary

gvilums replied:
The `cwd` member does seem to be used in the `kind` function [here](https://github.com/oven-sh/bun/blob/26a720defe3a0962f8b1062734b221b838613d4a/src/fs.zig#L1252-L1270), but that's the only spot. The other places in the `fs` implementation just use `std.fs.cwd`. So I think to ensure that the `kind` of a file is correctly resolved this should be updated (but maybe I'm wrong on that, in general the `cwd` handling is kind of a mess)

gvilums commented:
> 4 issues before this can be shipped.
> 
> 1. This approach is not fast enough. It needs to be an iterator so that the developer gets immediate feedback instead of waiting for potentially their entire directory tree to parse every package.json file.
This is now implemented in the `PackageFilterIterator ` in `filter_args.zig`
> 2. There is a resource leak. After enough scripts are run, spawning any process fails with `SystemResources` which usually means a file descriptor leak.
I'm able to reproduce this, but running `lsof -p <pid>` to list the currently open file descriptors shows only a handful at any given time. Not sure what's causing this issue. 
> 3. it scans node_modules. It should not scan node_modules, .git, or any `.` directories.
Implemented by passing an ignore function to the `GlobWalker`
> 4. We need tests for what happens when you run multiple JavaScript/TypeScript files in different projects same process (`bun run --filter="*" index.js`). Currently, Bun assumes only one JavaScript file every executes for a given thread. That assumption is not true when using these code paths.
We now `fork` before executing any JS. This of course doesn't work on windows, and more generally we should decide how to handle this. Options include:
- Figure out the correct command line arguments and spawn a new copy of the bun process. Slowest, but would provide complete isolation. 
- Continue using fork and use something like `RtlCloneUserProcess` on windows. Caveat is that these are not "official" windows APIs and their stability is unknown.
- Work on isolating VM instances and run everything in the same process. Fastest, but also potentially risky?
> 5. It would be best to use the module resolution cache so that when running a JavaScript file later, the subsequent calls to getdents64() get cached. This will involve a small change to the GlobWalker.
Implemented by abstracting how the `GlobWalker` does filesystem access.

gvilums commented:
After investigating 2. more, it seems like the error lies somewhere in child_process.spawnAndWait. Specifically, the error that's happening is that the child fails between being forked and calling execv, and signals this error to the parent.

MickL commented:
Does this also work with `bun add`? It is very annoying that I always need to go in the directory. pnpn has `pnp add --filter workspace`

dylan-conway commented on src/glob.zig:
```diff
@@ -624,6 +784,18 @@ pub fn GlobWalker_(
             only_files: bool,
         ) !Maybe(void) {
             errdefer arena.deinit();
+            var cwd: []const u8 = undefined;
```
Is this `cwd` supposed to be used with `initWithCwd` on line 788?

gvilums replied:
I think this is leftover code, I removed the cwd stuff

dylan-conway commented on src/deps/tinycc:
It looks like there isn't a diff, was this change intentional?

gvilums replied:
nope, not intended

Jarred-Sumner commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,302 @@
+        defer allocator.free(json_buf);
+        const json_len = try json_file.preadAll(json_buf, 0);
+        const json_source = bun.logger.Source.initPathString(json_path, json_buf[0..json_len]);
+        const json = try json_parser.ParseJSONUTF8(&json_source, log, allocator);
```
We are parsing the package.json files multiple times here since it may happen again later 

I'm hesitant to block the PR on this but we should figure out a way to cache this to avoid the overhead 

Also, we can simplify this code a bit:

bun.sys.File.toSource(json_path, allocator);

Jarred-Sumner commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,302 @@
+                const parts = [_]string{filter_utf8};
+                filter_utf8 = bun.path.joinAbsStringBuf(cwd, &buf, &parts, .auto);
+            }
+            var filter_utf32 = std.ArrayListUnmanaged(u32){};
```
feels like this will be very slow but i guess the filters won't be long so it should be fine

Jarred-Sumner commented on src/cli/run_command.zig:
```diff
@@ -348,13 +349,23 @@ pub const RunCommand = struct {
             .argv = &argv,
             .argv0 = shell_bin.ptr,
 
             // TODO: remember to free this when we add --filter or --concurrent
```
```suggestion
            // TODO: remember to free this when we add --concurrent
```

paperdave replied:
envp is freed. this entire comment can be removed now.

dylan-conway commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,302 @@
+        return true;
+    }
+    // skip node_modules
+    if (strings.eqlComptime(val, "node_modules")) {
```
we could skip `CMakeFiles` as well, but it's less likely to exist so probably not necessary

Jarred-Sumner replied:
we should also skip`.git`

Jarred-Sumner replied:
but that can be added later

dylan-conway commented on src/cli/filter_arg.zig:
```diff
@@ -0,0 +1,302 @@
+                    try out_patterns.append(pattern);
+                },
+                else => {
+                    // TODO log error and fail
```
We need an error message here

gvilums replied:
Added

paperdave commented on src/cli/run_command.zig:
```diff
@@ -348,13 +349,23 @@ pub const RunCommand = struct {
             Output.flush();
         }
 
+        const envp = try env.map.createNullDelimitedEnvMap(allocator);
```
on a different branch i had a version of this function that did one large allocation so that the defer free would be a single free.

code is https://github.com/oven-sh/bun/pull/9642 and might be cleaner to use.

it was put on pause because of change of focus to get 1.1 out.

paperdave commented on src/cli/bunx_command.zig:
```diff
@@ -525,8 +525,7 @@ pub const BunxCommand = struct {
                 );
-                // runBinary is noreturn
-                @compileError("unreachable");
+                Global.exit(if (res.notFailure()) 0 else 1);
```
this is expected to exit with the code the subprocess does,
and if the subprocess exits with a signal code instead of an exit code (SIGSEGV, SIGILL), we should raise the same one.

i hope there are already tests for this but you can try to `bunx node -e 'process.abort()'`

gvilums replied:
Makes sense, I updated the ExecResult to include an `exit` method which does the right thing

Jarred-Sumner reviewed (commented):
this needs more tests:
- what happens when a permissions error occurs when reading package.json or globbing? 
- what happens when a package.json is malformed?
- we should have a test that bun returns the appropriate exit code when a script run via --filter exits with a non-zero exit code
- we should look at tests from pnpm or yarn or npm to copy from
- a test that checks script & package names with non-ascii characters works

paperdave commented on src/cli/run_command.zig:
```diff
@@ -1280,13 +1290,144 @@ pub const RunCommand = struct {
         Output.flush();
     }
 
+    pub const ExecResult = union(enum) {
```
can you add `signal: u32` here, this will make this similar to `std.process.Child.Term`. stopped and unknown are not as important.

and then i recommend adding some function in here to exit with the execresult, which will call exitWide for the codes, but use `raiseIgnoringPanicHandler` for signals. (posix only)

gvilums replied:
Done

dylan-conway commented:
> this needs more tests:
> 
> * what happens when a package.json is malformed?

Missing name and empty name are two tests we should have. Pnpm seems to not allow filtering with `--filter ""` so we can exit early when that pattern is given

Jarred-Sumner commented:
> Missing name and empty name are two tests we should have. 

It's reading package.jsons to check the name/version in this code, so we need to test that here too

beorn commented:
Perhaps also support the '-F' shorthand?

kevinmarrec commented:
@gvilums @Jarred-Sumner @dylan-conway 

Dependency order is not respected as said in the docs since this PR
https://github.com/oven-sh/bun/issues/12203
https://github.com/oven-sh/bun/issues/13239

I may have never worked, or there has been a breaking chance since.
